data/archiver/stream.py

```diff
@@ -126,46 +131,88 @@ def stream_with_compression_and_md5_using_tmpfile(self, ftp, fin, fout, cb):
     def is_compressed(self, s3url):
         return s3url.endswith('.gz') and self.s3.read_block(s3url, 0, 2) == b'\x1f\x8b'
 
-    def start(self, res: DataArchiverRequest):
+    def start(self, res: DataArchiverResult):
         total_size = 0
+        prefix_paths = {}
         for file in res.files:
             if not file.success:
                 continue
 
             if self.s3.exists(file.cloud_url):
+                url = S3Url(file.cloud_url)
+                environment = url.bucket.split('-')[-1]
+                prefix_paths.setdefault(environment, {})[url.uuid] = f'{environment}/{url.uuid}'
+
                 size = self.s3.size(file.cloud_url)
                 total_size += size
                 file.size = size
             else:
                 file.error = 'File not found in S3.'
                 file.success = False
-        
+
         total_files = len(res.files)
-        num_files = sum(map(lambda f : f.success, res.files))
-        
+        num_files = sum(map(lambda f: f.success, res.files))
         if total_files != num_files:
             self.logger.info(f'{total_files - num_files} files not found.')
 
-        pbar = tqdm(total=total_size, unit='B', unit_scale=True, desc=f'{num_files} files', mininterval=2)
-        pool = ThreadPool() # cpu_count() DEFAULT_THREAD_COUNT=25
-        def cp(file):
-
-            if not file.success:
+        # Create each environment/uuid directory once for all files
+        # rather than have all files try to create it
+        for environment, uuids in prefix_paths.items():
```

Comment:
I don't really understand this loop as we only archive/send data to only 1 environment (dev/staging/prod) in an archiving process. Also, 1 archiving process only belongs to 1 submission. That means there should be only 1 submission uuid.

Reply:
I don't think it's work-in-progress, but it is part of an investigation.

If everything is working well this code will:
- change to the environment directory (creating it if needed)
- make a submission-uuid directory inside the environment directory ( or do nothing if one already exists)

Because data exports are just collections of cloud urls, which is what is being used to pickup the environment and submission uuid:
- s3 cloud urls should all end with the same environment, but it might not
- the cloud url uuid should match the submission uuid id, but they might not

So I'm coding defensively around what what is possible to occur rather than what is designed to occur, apologies if that is hard to understand.